### PR TITLE
Move `DatumVec` to `repr`

### PR DESCRIPTION
**Pull request by frankmcsherry**
This PR moves `render::DatumVec` to `repr`, so that it can be more broadly used to avoid `Row::unpack` calls. Many of these calls are also removed (but not all of them, for .. reasons).

### Tips for reviewer

Each of the modifications are meant to not change an observed behavior, and just mean to reduce allocations.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/coord/src/coord.rs [hunk 1]
```diff
@@ -3648,13 +3648,14 @@ where
                         |rows: Vec<Row>| -> Result<Vec<(Row, Diff)>, CoordError> {
                             // Use 2x row len incase there's some assignments.
                             let mut diffs = Vec::with_capacity(rows.len() * 2);
+                            let mut datum_vec = repr::DatumVec::new();
```
Comment (quodlibetor) [inline src/coord/src/coord.rs:3651]: is it possible and worth it to move this datum_vec up to where we create the arena on line 3644?

File: src/dataflow/src/logging/differential.rs [hunk 2]
```diff
@@ -26,9 +26,9 @@ use crate::activator::RcActivator;
 use crate::arrangement::manager::RowSpine;
 use crate::arrangement::KeysValsHandle;
 use crate::logging::ConsolidateBuffer;
-use crate::render::datum_vec::DatumVec;
 use crate::render::Permutation;
 use crate::replay::MzReplay;
+use repr::DatumVec;
 use repr::{Datum, Row, Timestamp};
```
Comment (quodlibetor) [inline src/dataflow/src/logging/differential.rs:32]: nit (that should have been automatically handled by tooling?) these two lines should be merged.

```suggestion
use repr::{Datum, DatumVec, Row, Timestamp};
```

File: src/dataflow/src/logging/materialized.rs [hunk 3]
```diff
@@ -25,11 +25,11 @@ use super::{LogVariant, MaterializedLog};
 use crate::activator::RcActivator;
 use crate::arrangement::manager::RowSpine;
 use crate::arrangement::KeysValsHandle;
-use crate::render::datum_vec::DatumVec;
 use crate::render::Permutation;
 use crate::replay::MzReplay;
 use expr::{GlobalId, SourceInstanceId};
 use repr::adt::jsonb::Jsonb;
+use repr::DatumVec;
 use repr::{Datum, Row, Timestamp};
```
Comment (quodlibetor) [inline src/dataflow/src/logging/materialized.rs:33]: same as above

```suggestion
use repr::{Datum, DatumVec, Row, Timestamp};
```

File: src/dataflow/src/logging/timely.rs [hunk 4]
```diff
@@ -24,10 +24,10 @@ use crate::activator::RcActivator;
 use crate::arrangement::manager::RowSpine;
 use crate::arrangement::KeysValsHandle;
 use crate::logging::ConsolidateBuffer;
-use crate::render::datum_vec::DatumVec;
 use crate::render::Permutation;
 use crate::replay::MzReplay;
 use dataflow_types::logging::LoggingConfig;
+use repr::DatumVec;
 use repr::{datum_list_size, datum_size, Datum, Row, Timestamp};
```
Comment (quodlibetor) [inline src/dataflow/src/logging/timely.rs:31]: Similar to above:

```suggestion
use repr::{datum_list_size, datum_size, Datum, DatumVec, Row, Timestamp};
```

File: src/dataflow/src/render/context.rs [hunk 5]
```diff
@@ -29,10 +29,10 @@ use timely::progress::{Antichain, Timestamp};
 
 use crate::arrangement::manager::{ErrSpine, RowSpine, TraceErrHandle, TraceRowHandle};
 use crate::operator::CollectionExt;
-use crate::render::datum_vec::DatumVec;
 use crate::render::Permutation;
 use dataflow_types::{DataflowDescription, DataflowError};
 use expr::{GlobalId, Id, MapFilterProject, MirScalarExpr};
+use repr::DatumVec;
 use repr::{Diff, Row, RowArena};
```
Comment (quodlibetor) [inline src/dataflow/src/render/context.rs:36]: same

```suggestion
use repr::{DatumVec, Diff, Row, RowArena};
```

File: src/dataflow/src/render/sinks.rs [hunk 6]
```diff
@@ -113,20 +113,26 @@ where
 
         let keyed = if user_key_indices.is_some() {
             let key_indices = user_key_indices.expect("known to exist");
+            let mut datum_vec = repr::DatumVec::new();
             collection.map(move |row| {
                 // TODO[perf] (btv) - is there a way to avoid unpacking and repacking every row and cloning the datums?
                 // Does it matter?
```
Comment (quodlibetor) [inline src/dataflow/src/render/sinks.rs:119]: is this comment (and the next one in this file) now obsolete/need updating?

File: src/coord/src/coord.rs (hunk 1 above)
Comment (frankmcsherry) [reply src/coord/src/coord.rs:3651]: I put it here on the theory that .. in other cases we don't need it, and afaict it isn't any different here vs there. I could be wrong about that, though! But, I didn't know how often we are in this one case vs the others, is why I did it this way.

File: src/dataflow/src/render/sinks.rs (hunk 6 above)
Comment (frankmcsherry) [reply src/dataflow/src/render/sinks.rs:119]: I think it is still as valid. The "unpacking" still happens, but it doesn't allocate any more.

**Comment by frankmcsherry** [comment]
So there does seem to be some change in observed behavior: arbitrary row sorting seems to be different now (though, possibly better). In particular, I think ties are now being broken by `Datum` interpretation rather than by `Row` interpretation. Can probably do it the old way too, though.

**Review by antiguru — APPROVED** [review]
LGTM, thanks for this change!

**Pull request merged**